### PR TITLE
FM-519 Buyer Detail redirects if invalid buyer details

### PR DESCRIPTION
**Pull request by ManFromMons**
Schema changes from running db:migrate - please commit these
Logic in the framework controller in beta - added a before_action to validate buyer_detail...

Rails folk - please validate if this is the right strategy - it seems reasonable from where I am sitting....

Starting a quick-search doesn't go to Beta so this code won't have any effect and the user will be be able to start a procurement.  This logic will engage when the user enters the beta journey - at the Search Results page.  So they will be directed to buyer-details when they have done 2 pages of a quick search.  Buyer-details doesn't know how to direct the user back again.

Added logic to enable the buyer-detail redirect to pass through the originating querystring and url - the buyer-detail controller now redirects to those details if supplied.  This will stop the new procurement journey being broken.

### CODE REVIEW

File: app/controllers/facilities_management/beta/framework_controller.rb [hunk 1]
```diff
@@ -3,12 +3,28 @@ module Beta
...
       def authorize_user
         authorize! :read, FacilitiesManagement
       end
+
+      def capture_buyer_detail
+        if guarded_path? && current_user&.fm_buyer_details_incomplete? &&
+           request.fullpath != edit_facilities_management_beta_buyer_detail_path(FacilitiesManagement::BuyerDetail.find_or_create_by(user: current_user))
+          redirect_to(edit_facilities_management_beta_buyer_detail_path(FacilitiesManagement::BuyerDetail.find_or_create_by(user: current_user)))
+        end
+      end
```
Comment (doctorpod) [inline app/controllers/facilities_management/beta/framework_controller.rb]:
```suggestion
      def capture_buyer_detail
        edit_path = edit_facilities_management_beta_buyer_detail_path(FacilitiesManagement::BuyerDetail.find_or_create_by(user: current_user))
        if guarded_path? && current_user&.fm_buyer_details_incomplete? &&
           request.fullpath != edit_path
          redirect_to(edit_path)
        end
      end
```

Comment (doctorpod) [reply app/controllers/facilities_management/beta/framework_controller.rb]: Even like this, methods with side effects always make me nervous. Maybe change the name of the method to reflect that we have potentially added a BuyerDetail.

**Comment by kennyevil** [comment]
This is blowing up a load of controller tests. I'm assuming because the factory user doesn't have a buyer detail generated for it.

**Comment by ManFromMons** [comment]
11 Large impact test failures occur in the feature tests around procurements.
I will add some tests to prove this feature works.
And rework the failing tests - likely by ensuring a buyer_detail exists for the tests

**Comment by ManFromMons** [comment]
> This is blowing up a load of controller tests. I'm assuming because the factory user doesn't have a buyer detail generated for it.

Correct - the problem resolved in subsequent commits

**Pull request merged**